### PR TITLE
[minor] Fixes for oath: Set the rfc=true for the curve values, and allow priming the updaters so the keys are available immediately.

### PR DESCRIPTION
**Pull request by scr-oath**
# Description

- Oauth2 tokens signed in Oath/Yahoo/VerizonMediaGroup describe the curve in more detail; including the `rfc=true` query parameter is required to ensure they are returned in something that [jwx](https://github.com/lestrrat-go/jwx) understands
   - ~Added `WithAthenzRFC` option to `authorizer` and `WithAthenzJwksURL` to `jwkd`~
- ~Update jwx to the latest version~
- ~Allow options to "prime" (update before starting) the various daemons so that, for instance, all the keys and policies are available immediately.~
- The authorized roles are needed, so add a variant to the policyd: CheckPolicyRoles, which returns a slice of authorized roles when `err == nil`
- For testing, brought in [testify](https://github.com/stretchr/testify) for assertions and deep comparison.

## Type of change

- [X] Bug fix
- [X] New feature
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Non-code changes (update documentation, pipeline, etc.)

### Flags

- [ ] breaks backward compatibility
- [X] requires a documentation update
- [ ] has untestable code

---

## Checklist

- [x] Followed the guidelines in the CONTRIBUTING document
- [x] Added prefix `[major]`/`[minor]`/`[patch]`/`[skip]` in the PR title
- [x] Tested and linted the code
- [x] Commented the code
- [x] Made corresponding changes to the documentation
- [x] Confirmed no dropping in test coverage (by [Codecov](https://codecov.io/gh/yahoojapan/athenz-authorizer/pulls))
- [x] Passed all pipeline checking
- [x] Approved by >1 reviewer

## Checklist for maintainer
- [x] Use `Squash and merge`
- [x] Double-confirm the merge message has prefix `[major]`/`[minor]`/`[patch]`/`[skip]`
- [x] Delete the branch after merge

### CODE REVIEW

**Comment by WindzCUHK** [comment]
The new jwx library require go 1.15. The CI pipeline is using 1.14 currently, which will cause the pipeline to fail.

Can you help to update the following go versions? 1.15 is good, but 1.16 is preferred.
- https://github.com/yahoojapan/athenz-authorizer/blob/d1054bbae8ac05316568dec5f53a099e47a947ed/go.mod#L3
- https://github.com/yahoojapan/athenz-authorizer/blob/d1054bbae8ac05316568dec5f53a099e47a947ed/.circleci/config.yml#L5
- https://github.com/yahoojapan/athenz-authorizer/blame/master/policy/CONTRIBUTING.md#L43

On 1.16, I got the following error after running the unit test.
Can you help to have a check?
```console
$ go test --race ./... > out.txt
# github.com/yahoojapan/athenz-authorizer/v5/jwk [github.com/yahoojapan/athenz-authorizer/v5/jwk.test]
jwk/daemon_test.go:232:22: k1.(*jwk.Set).Keys undefined (type *jwk.Set is pointer to interface, not interface)
jwk/daemon_test.go:232:55: k2.(*jwk.Set).Keys undefined (type *jwk.Set is pointer to interface, not interface)
jwk/daemon_test.go:233:112: k1.(*jwk.Set).Keys undefined (type *jwk.Set is pointer to interface, not interface)
jwk/daemon_test.go:233:143: k2.(*jwk.Set).Keys undefined (type *jwk.Set is pointer to interface, not interface)
jwk/daemon_test.go:367:27: val.(*jwk.Set).Keys undefined (type *jwk.Set is pointer to interface, not interface)
jwk/daemon_test.go:408:27: val.(*jwk.Set).Keys undefined (type *jwk.Set is pointer to interface, not interface)
jwk/daemon_test.go:419:26: val.(*jwk.Set).Keys undefined (type *jwk.Set is pointer to interface, not interface)
jwk/daemon_test.go:508:27: val.(*jwk.Set).Keys undefined (type *jwk.Set is pointer to interface, not interface)
jwk/daemon_test.go:572:29: val1.(*jwk.Set).Keys undefined (type *jwk.Set is pointer to interface, not interface)
jwk/daemon_test.go:581:29: val2.(*jwk.Set).Keys undefined (type *jwk.Set is pointer to interface, not interface)
jwk/daemon_test.go:581:29: too many errors

$ go version
go version go1.16.6 darwin/amd64
```

**Comment by scr-oath** [comment]
> The new `jwx` library requires go 1.15. The CI pipeline is using 1.14 currently, which will cause the pipeline to fail.

Yeah, was able to revert to older version and make it work merely with the `rpc=true` query param; #89 upgrades go independently, and we can follow _that_ one with upgrading jwk so three separate changes

**Comment by WindzCUHK** [comment]
@scr-oath 

1. Is there any reason for adding the `prime*d` options? As the `*d.Start()` functions are async, there is chance that when the `func (a *authority) Start(ctx context.Context)` returns, nothing is fetched inside authorizerd which causes all requests to be denied.
https://github.com/yahoojapan/athenz-authorizer/blob/9829eb43873f7327f99a795876e57f7f4780fd1c/authorizerd.go#L78
2. Can you help to add new test cases for checking the below code section in `authorizerd_test.go`? The old test cases did not verify the newly added `authorizedRoles` field.
https://github.com/yahoojapan/athenz-authorizer/blob/9829eb43873f7327f99a795876e57f7f4780fd1c/authorizerd.go#L466-L471

**Comment by scr-oath** [comment]
@WindzCUHK 
> 1. Is there any reason for adding the `prime*d` options? As the `*d.Start()` functions are async, there is chance that when the `func (a *authority) Start(ctx context.Context)` returns, nothing is fetched inside authorizerd which causes all requests to be denied.

authorizerd.Start is not 100% async - it returns a channel after starting the other daemons.  The prime*d options are intended to force a _synchronous_ call to Update in the synchronous part of authorizerd.Start and, since they are updated, then started in the right dependency sequence, prime, specifically avoids the case that when it returns the authorizer cannot be used - based on the defaults, it was my understanding that, in fact, without priming to force an update, the entire application could not authorize anything for a full hour, before the update was triggered.  If there's some other mechanism to ensure that the authorizerd is ready to use by waiting for something (that doesn't take a full hour), then let me know, but I thought that the forced call to Update made it useable (at least in the happy path, short of any errors) immediately

>    https://github.com/yahoojapan/athenz-authorizer/blob/9829eb43873f7327f99a795876e57f7f4780fd1c/authorizerd.go#L78
> 2. Can you help to add new test cases for checking the below code section in `authorizerd_test.go`? The old test cases did not verify the newly added `authorizedRoles` field.
>    https://github.com/yahoojapan/athenz-authorizer/blob/9829eb43873f7327f99a795876e57f7f4780fd1c/authorizerd.go#L466-L471

Sure, I can take a look

**Comment by WindzCUHK** [comment]
@scr-oath 

> If there's some other mechanism to ensure that the authorizerd is ready to use by waiting for something (that doesn't take a full hour), then let me know, but I thought that the forced call to Update made it useable
Can the `Init()` function help?
- update README in https://github.com/yahoojapan/athenz-authorizer/pull/90
- example
    - https://github.com/yahoojapan/authorization-proxy/blob/15fed79e58e4da58276aa7f01ee6aebab5214692/main.go#L108-L112
    - https://github.com/yahoojapan/authorization-proxy/blob/15fed79e58e4da58276aa7f01ee6aebab5214692/usecase/authz_proxyd.go#L71-L87
    - https://github.com/yahoojapan/authorization-proxy/blob/15fed79e58e4da58276aa7f01ee6aebab5214692/service/authorizerd.go#L23

**Comment by scr-oath** [comment]
> @scr-oath
> 
> > If there's some other mechanism to ensure that the authorizerd is ready to use by waiting for something (that doesn't take a full hour), then let me know, but I thought that the forced call to Update made it useable
> 
> Can the `Init()` function help?

Yep - I just tried it, that was just the trick…

**Review by WindzCUHK — APPROVED** [review]
LGTM

File: policy/daemon.go [hunk 1]
```diff
@@ -223,39 +238,39 @@ func (p *policyd) CheckPolicy(ctx context.Context, domain string, roles []string
...
-			glg.Debugf("check policy domain: %s, role: %v, action: %s, resource: %s, result: %v", domain, roles, action, resource, err)
-			return err
+	var allowedRoles []string
+	for re := range ech {
+		if re.Effect != nil { // denied assertion is prioritize, so return directly
+			glg.Debugf("check policy domain: %s, role: %v, action: %s, resource: %s, result: %v", domain, roles, action, resource, re.Effect)
+			return nil, re.Effect
 		}
-		allowed = true
+		allowedRoles = append(allowedRoles, re.Role)
```
Comment (scr-oath) [inline policy/daemon.go:265]: Would be nice if the allowedRoles were allocated big enough to hold all results to prevent re-allocation

Comment (scr-oath) [reply policy/daemon.go:265]: Done

**Pull request merged**